src/details/ArborX_DetailsDendrogram.hpp

```diff
@@ -89,33 +88,29 @@ template <typename ExecutionSpace, typename MemorySpace>
 void dendrogramUnionFind(ExecutionSpace const &exec_space,
                          Kokkos::View<WeightedEdge const *, MemorySpace> edges,
                          Kokkos::View<int *, MemorySpace> &parents,
-                         Kokkos::View<float *, MemorySpace> &parent_heights)
+                         Kokkos::View<float *, MemorySpace> &weights)
 {
   Kokkos::Profiling::pushRegion("ArborX::Dendrogram::dendrogram_union_find");
 
   auto const num_edges = edges.size();
   auto const num_vertices = num_edges + 1;
 
-  KokkosExt::reallocWithoutInitializing(exec_space, parent_heights, num_edges);
-
-  Kokkos::Profiling::pushRegion("ArborX::Dendrogram::edge_sort");
-  Kokkos::parallel_for(
-      "ArborX::Dendrogram::copy_weights",
-      Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, num_edges),
-      KOKKOS_LAMBDA(int const e) { parent_heights(e) = edges(e).weight; });
-  auto permute = sortObjects(exec_space, parent_heights);
-  Kokkos::View<UnweightedEdge *, MemorySpace> sorted_unweighted_edges(
+  KokkosExt::reallocWithoutInitializing(exec_space, weights, num_edges);
+  Kokkos::View<UnweightedEdge *, MemorySpace> unweighted_edges(
       Kokkos::view_alloc(
           exec_space, Kokkos::WithoutInitializing,
           "ArborX::Dendrogram::dendrogram_union_find::unweighted_edges"),
       num_edges);
   Kokkos::parallel_for(
-      "ArborX::Dendrogram::dendrogram_union_find::permute_and_fill",
+      "ArborX::Dendrogram::copy_weights_and_edges",
       Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, num_edges),
-      KOKKOS_LAMBDA(int i) {
-        auto &edge = edges(permute(i));
-        sorted_unweighted_edges(i) = {edge.source, edge.target};
+      KOKKOS_LAMBDA(int const e) {
+        weights(e) = edges(e).weight;
+        unweighted_edges(e) = {edges(e).source, edges(e).target};
       });
+
+  Kokkos::Profiling::pushRegion("ArborX::Dendrogram::edge_sort");
+  KokkosExt::sortByKey(exec_space, weights, unweighted_edges);
```

Comment:
While it is less code, you do pay in performance as there is a hidden cost here. During the `sortByKey` you actually use more data. Instead of two arrays, each with value type of size 4 bytes, you are doing two arrays with one of size 4 and one of size 8. So the sort itself could be slower. Have you measured that it is actually faster?

Reply:
**CURRENT MASTER**
```
ArborX version    : 1.4 (dev)
ArborX hash       : 73fb555
Kokkos version    : 3.7.99
algorithm         : hdbscan
minpts            : 2
filename          : /gpfs/alpine/proj-shared/csc333/hacc_37M.arborx [binary, max_pts = -1]
samples           : -1
verbose           : false
Reading in "/gpfs/alpine/proj-shared/csc333/hacc_37M.arborx" in binary mode...done
Read in 36902719 3D points

BEGIN KOKKOS PROFILING REPORT:
TOTAL TIME: 3.5505 seconds
TOP-DOWN TIME TREE:
<average time> <percent of total time> <percent time in Kokkos> <percent MPI imbalance> <remainder> <kernels per second> <number of calls> <name> [type]
===================
|-> 3.22e+00 sec 90.8% 30.3% 0.0% 0.0% 5.71e+01 1 ArborX::HDBSCAN::total [region]
|   |-> 3.22e+00 sec 90.8% 30.3% 0.0% 0.0% 5.68e+01 1 ArborX::HDBSCAN [region]
|   |   |-> 2.37e+00 sec 66.7% 5.9% 0.0% 0.0% 4.65e+00 1 ArborX::HDBSCAN::dendrogram [region]
|   |   |   |-> 2.37e+00 sec 66.7% 5.9% 0.0% 0.1% 4.65e+00 1 ArborX::Dendrogram::Dendrogram [region]
|   |   |       |-> 2.37e+00 sec 66.6% 5.9% 0.0% 0.0% 3.80e+00 1 ArborX::Dendrogram::dendrogram_union_find [region]
|   |   |       |   |-> 2.26e+00 sec 63.5% 1.8% 0.0% 0.0% 1.33e+00 1 ArborX::Dendrogram::dendrogram_union_find::union_find [region]
|   |   |       |   |   |-> 2.26e+00 sec 63.5% 1.8% 0.0% 0.0% 1.33e+00 1 ArborX::Dendrogram::dendrogram_union_find::union_find_host [region]
|   |   |       |   |       |-> 2.23e+00 sec 62.8% 0.0% 0.0% 100.0% 0.00e+00 1 ArborX::Dendrogram::dendrogram_union_find::union_find [region]
|   |   |       |   |       |-> 1.38e-02 sec 0.4% 200.0% 0.0% ------ 1 "ArborX::Dendrogram::labels"="(none)" [copy]
|   |   |       |   |       |   |-> 1.38e-02 sec 0.4% 100.0% 0.0% ------ 1 Kokkos::ViewFill-1D [for]
|   |   |       |   |       |-> 1.34e-02 sec 0.4% 100.0% 0.0% ------ 1 ArborX::Algorithms::iota [for]
|   |   |       |   |-> 6.93e-02 sec 2.0% 99.9% 0.0% 0.1% 2.89e+01 1 ArborX::Dendrogram::dendrogram_union_find::copy_to_host [region]
|   |   |       |   |   |-> 4.38e-02 sec 1.2% 100.0% 0.0% ------ 1 "ArborX::Dendrogram::dendrogram_union_find::unweighted_edges"="ArborX::Dendrogram::dendrogram_union_find::unweighted_edges" [copy]
|   |   |       |   |   |-> 2.54e-02 sec 0.7% 100.0% 0.0% ------ 1 Kokkos::View::initialization [ArborX::Dendrogram::parents_mirror] via memset [for]
|   |   |       |   |-> 2.48e-02 sec 0.7% 100.0% 0.0% 0.0% 4.03e+01 1 ArborX::Dendrogram::dendrogram_union_find::copy_to_device [region]
|   |   |       |   |   |-> 2.48e-02 sec 0.7% 100.0% 0.0% ------ 1 "ArborX::Dendrogram::parents"="ArborX::Dendrogram::parents_mirror" [copy]
|   |   |       |   |-> 1.56e-02 sec 0.4% 35.1% 0.0% 0.6% 1.93e+02 1 ArborX::Dendrogram::edge_sort [region]
|   |   |       |       |-> 1.02e-02 sec 0.3% 2.1% 0.0% 97.9% 9.77e+01 1 ArborX::Sorting [region]
|   |   |       |       |-> 4.75e-03 sec 0.1% 100.0% 0.0% ------ 1 ArborX::Dendrogram::dendrogram_union_find::permute_and_fill [for]
```
**THIS PR**
```
ArborX version    : 1.4 (dev)
ArborX hash       : 14d817c
Kokkos version    : 3.7.99
algorithm         : hdbscan
minpts            : 2
filename          : /gpfs/alpine/proj-shared/csc333/hacc_37M.arborx [binary, max_pts = -1]
samples           : -1
verbose           : false
^[[OReading in "/gpfs/alpine/proj-shared/csc333/hacc_37M.arborx" in binary mode...done
Read in 36902719 3D points

BEGIN KOKKOS PROFILING REPORT:
TOTAL TIME: 3.43455 seconds
TOP-DOWN TIME TREE:
<average time> <percent of total time> <percent time in Kokkos> <percent MPI imbalance> <remainder> <kernels per second> <number of calls> <name> [type]
===================
|-> 3.21e+00 sec 93.4% 30.2% 0.0% 0.0% 5.67e+01 1 ArborX::HDBSCAN::total [region]
|   |-> 3.21e+00 sec 93.4% 30.2% 0.0% 0.0% 5.64e+01 1 ArborX::HDBSCAN [region]
|   |   |-> 2.35e+00 sec 68.5% 5.7% 0.0% 0.0% 3.82e+00 1 ArborX::HDBSCAN::dendrogram [region]
|   |   |   |-> 2.35e+00 sec 68.5% 5.7% 0.0% 0.0% 3.82e+00 1 ArborX::Dendrogram::Dendrogram [region]
|   |   |       |-> 2.35e+00 sec 68.5% 5.7% 0.0% 0.0% 2.97e+00 1 ArborX::Dendrogram::dendrogram_union_find [region]
|   |   |       |   |-> 2.25e+00 sec 65.4% 1.8% 0.0% 0.0% 1.33e+00 1 ArborX::Dendrogram::dendrogram_union_find::union_find [region]
|   |   |       |   |   |-> 2.25e+00 sec 65.4% 1.8% 0.0% 0.0% 1.34e+00 1 ArborX::Dendrogram::dendrogram_union_find::union_find_host [region]
|   |   |       |   |       |-> 2.22e+00 sec 64.6% 0.0% 0.0% 100.0% 0.00e+00 1 ArborX::Dendrogram::dendrogram_union_find::union_find [region]
|   |   |       |   |       |-> 1.38e-02 sec 0.4% 200.0% 0.0% ------ 1 "ArborX::Dendrogram::labels"="(none)" [copy]
|   |   |       |   |       |   |-> 1.38e-02 sec 0.4% 100.0% 0.0% ------ 1 Kokkos::ViewFill-1D [for]
|   |   |       |   |       |-> 1.38e-02 sec 0.4% 100.0% 0.0% ------ 1 ArborX::Algorithms::iota [for]
|   |   |       |   |-> 6.93e-02 sec 2.0% 99.9% 0.0% 0.1% 2.88e+01 1 ArborX::Dendrogram::dendrogram_union_find::copy_to_host [region]
|   |   |       |   |   |-> 4.39e-02 sec 1.3% 100.0% 0.0% ------ 1 "ArborX::Dendrogram::dendrogram_union_find::unweighted_edges"="ArborX::Dendrogram::dendrogram_union_find::unweighted_edges" [copy]
|   |   |       |   |   |-> 2.54e-02 sec 0.7% 100.0% 0.0% ------ 1 Kokkos::View::initialization [ArborX::Dendrogram::parents_mirror] via memset [for]
|   |   |       |   |-> 2.24e-02 sec 0.7% 100.0% 0.0% 0.0% 4.46e+01 1 ArborX::Dendrogram::dendrogram_union_find::copy_to_device [region]
|   |   |       |   |   |-> 2.24e-02 sec 0.7% 100.0% 0.0% ------ 1 "ArborX::Dendrogram::parents"="ArborX::Dendrogram::parents_mirror" [copy]
|   |   |       |   |-> 1.28e-02 sec 0.4% 0.0% 0.0% 100.0% 0.00e+00 1 ArborX::Dendrogram::edge_sort [region]

Reply:
In summary
MASTER = sort (1.02e-02) + copy_weight (???) + permute_and_fill (4.75e-03) + iota (???)
THIS PR = sort (1.27e-02) + copy_weights_and_edges (???)

Reply:
I ran on Mac in Serial, also is OK.